File: packages/sdk-node/tests/bot-detection.test.ts
import { describe, expect, it } from 'vitest';
import {
  detectBot,
  isDatacenterIp,
  isHeadlessUserAgent,
  isHostingOrg,
} from '../src/utils/bot-detection';

describe('bot-detection', () => {
  describe('isDatacenterIp', () => {
    it('flags known datacenter IPv4 addresses', () => {
      expect(isDatacenterIp('34.72.176.129')).toBe(true); // Google Cloud
      expect(isDatacenterIp('51.75.1.1')).toBe(true); // OVH
      expect(isDatacenterIp('5.9.1.1')).toBe(true); // Hetzner
      expect(isDatacenterIp('159.65.1.1')).toBe(true); // DigitalOcean
    });

    it('flags known datacenter IPv6 addresses', () => {
      expect(isDatacenterIp('2001:41d0:350:1400::1')).toBe(true); // OVH
      expect(isDatacenterIp('[2a01:4f8::1]')).toBe(true); // Hetzner, bracketed
      expect(isDatacenterIp('2a01:4ff::1')).toBe(true); // Hetzner /29 upper bound
      expect(isDatacenterIp('::ffff:34.72.176.129')).toBe(true); // IPv4-mapped IPv6
    });

    it('does not flag residential / unknown / garbage IPs', () => {
      expect(isDatacenterIp('8.8.8.8')).toBe(false);
      expect(isDatacenterIp('192.168.1.10')).toBe(false);
      expect(isDatacenterIp('unknown')).toBe(false);
      expect(isDatacenterIp('')).toBe(false);
      expect(isDatacenterIp(null)).toBe(false);
      expect(isDatacenterIp(undefined)).toBe(false);
      expect(isDatacenterIp('not-an-ip')).toBe(false);
      expect(isDatacenterIp('999.999.999.999')).toBe(false);
    });
  });

  describe('isHeadlessUserAgent', () => {
    it('flags obvious automation / headless markers', () => {
      expect(isHeadlessUserAgent('Mozilla/5.0 (X11) HeadlessChrome/120.0.0.0 Safari/537.36')).toBe(
        true
      );
      expect(isHeadlessUserAgent('python-requests/2.31.0')).toBe(true);
      expect(isHeadlessUserAgent('curl/8.4.0')).toBe(true);
    });

    it('does not flag a real Chrome UA', () => {
      expect(
        isHeadlessUserAgent(
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
        )
      ).toBe(false);
      expect(isHeadlessUserAgent('')).toBe(false);
      expect(isHeadlessUserAgent(null)).toBe(false);
      expect(isHeadlessUserAgent(undefined)).toBe(false);
    });
  });

  describe('isHostingOrg', () => {
    it('flags unambiguous hosting providers', () => {
      expect(isHostingOrg('Hetzner Online GmbH')).toBe(true);
      expect(isHostingOrg('DigitalOcean, LLC')).toBe(true);
      expect(isHostingOrg('Vultr Holdings, LLC')).toBe(true);
    });

    it('does NOT flag broad consumer brands or non-hosting orgs', () => {
      expect(isHostingOrg('Google Fiber Inc.')).toBe(false);
      expect(isHostingOrg('Google LLC')).toBe(false);
      expect(isHostingOrg('AMAZON-02')).toBe(false);
      expect(isHostingOrg('Orange S.A.')).toBe(false);
      expect(isHostingOrg('')).toBe(false);
      expect(isHostingOrg(null)).toBe(false);
      expect(isHostingOrg(undefined)).toBe(false);
    });
  });

  describe('detectBot', () => {
    it('flags a datacenter IP with reason datacenter_ip', () => {
      expect(detectBot({ ip: '34.72.176.129' })).toEqual({
        isBot: true,
        reason: 'datacenter_ip',
      });
    });

    it('flags a hosting ASN with reason hosting_asn', () => {
      expect(detectBot({ ip: '8.8.8.8', org: 'Hetzner Online GmbH' })).toEqual({
        isBot: true,
        reason: 'hosting_asn',
      });
    });

    it('flags a headless UA with reason headless_ua', () => {
      expect(
        detectBot({
          ip: '8.8.8.8',
          userAgent: 'HeadlessChrome/120.0.0.0',
        })
      ).toEqual({ isBot: true, reason: 'headless_ua' });
    });

    it('does not flag a clean residential connection', () => {
      expect(
        detectBot({
          ip: '8.8.8.8',
          org: 'Orange S.A.',
          userAgent:
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        })
      ).toEqual({ isBot: false, reason: null });
    });

    it('prioritizes datacenter_ip over other signals', () => {
      expect(
        detectBot({
          ip: '5.9.1.1',
          org: 'Hetzner Online GmbH',
          userAgent: 'curl/8.4.0',
        })
      ).toEqual({ isBot: true, reason: 'datacenter_ip' });
    });
  });
});

File: packages/sdk-go/bot_detection_test.go
package pocketping

import "testing"

const realChromeUA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"

func TestIsDatacenterIP(t *testing.T) {
	datacenter := []string{
		"34.72.176.129", "34.123.170.104", "104.197.69.115", // GCP
		"51.75.1.1",             // OVH
		"5.9.1.1",               // Hetzner
		"159.65.1.1",            // DigitalOcean
		"2001:41d0:350:1400::1", // OVH v6
		"[2a01:4f8::1]",         // Hetzner v6, bracketed
	}
	for _, ip := range datacenter {
		if !IsDatacenterIP(ip) {
			t.Errorf("IsDatacenterIP(%q) = false, want true", ip)
		}
	}

	residential := []string{"86.247.12.34", "unknown", "", "not-an-ip"}
	for _, ip := range residential {
		if IsDatacenterIP(ip) {
			t.Errorf("IsDatacenterIP(%q) = true, want false", ip)
		}
	}
}

func TestIsHeadlessUserAgent(t *testing.T) {
	if !IsHeadlessUserAgent("Mozilla/5.0 HeadlessChrome/120") {
		t.Error("expected HeadlessChrome to be flagged")
	}
	if !IsHeadlessUserAgent("python-requests/2.31") {
		t.Error("expected python-requests to be flagged")
	}
	if IsHeadlessUserAgent(realChromeUA) {
		t.Error("real Chrome UA should not be flagged")
	}
	if IsHeadlessUserAgent("") {
		t.Error("empty UA should not be flagged")
	}
}

func TestIsHostingOrg(t *testing.T) {
	for _, org := range []string{"Hetzner Online GmbH", "DigitalOcean, LLC", "Vultr Holdings"} {
		if !IsHostingOrg(org) {
			t.Errorf("IsHostingOrg(%q) = false, want true", org)
		}
	}
	// Broad consumer brands rely on CIDR, not org matching.
	for _, org := range []string{"Google Fiber Inc.", "Google LLC", "AMAZON-02", "Orange S.A.", ""} {
		if IsHostingOrg(org) {
			t.Errorf("IsHostingOrg(%q) = true, want false", org)
		}
	}
}

func TestDetectBot(t *testing.T) {
	cases := []struct {
		name    string
		signal  BotSignal
		wantBot bool
		reason  string
	}{
		{"datacenter ip + spoofed UA", BotSignal{IP: "34.72.176.129", UserAgent: realChromeUA, Org: "Google LLC"}, true, "datacenter_ip"},
		{"hosting ASN, unlisted IP", BotSignal{IP: "203.0.113.7", UserAgent: realChromeUA, Org: "Vultr Holdings"}, true, "hosting_asn"},
		{"headless UA", BotSignal{IP: "86.247.12.34", UserAgent: "HeadlessChrome/120", Org: "Orange"}, true, "headless_ua"},
		{"real residential visitor", BotSignal{IP: "86.247.12.34", UserAgent: realChromeUA, Org: "Orange S.A."}, false, ""},
	}
	for _, c := range cases {
		got := DetectBot(c.signal)
		if got.IsBot != c.wantBot || got.Reason != c.reason {
			t.Errorf("%s: DetectBot = {%v %q}, want {%v %q}", c.name, got.IsBot, got.Reason, c.wantBot, c.reason)
		}
	}
}
